video/out/vo_dmabuf_wayland.c

```diff
@@ -0,0 +1,500 @@
+    } else if (CHECK_VA_STATUS(entry->vo, "vaExportSurfaceHandle()")) {
+        entry->drm_format = desc.layers[layer_no].drm_format;
+        for (int plane_no = 0; plane_no < desc.layers[layer_no].num_planes; ++plane_no) {
+            int object = desc.layers[layer_no].object_index[plane_no];
+            uint64_t modifier = desc.objects[object].drm_format_modifier;
+            zwp_linux_buffer_params_v1_add(params, desc.objects[object].fd, plane_no, desc.layers[layer_no].offset[plane_no], desc.layers[layer_no].pitch[plane_no], modifier >> 32, modifier & 0xffffffff);
+        }
+        dmabuf_imported = true;
+    }
+
+    /* clean up descriptor */
+    for (int i = 0; i < desc.num_objects; i++) {
+        close(desc.objects[i].fd);
+        desc.objects[i].fd = 0;
+    }
+
+    return dmabuf_imported;
+}
+
+static uintptr_t drmprime_key_provider(struct mp_image *src){
+    return (uintptr_t)src->planes[0];
+}
+
+static bool drmprime_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_entry *entry, struct zwp_linux_buffer_params_v1 *params)
+{
+    int layer_no, plane_no;
+    const AVDRMFrameDescriptor *avdesc = (AVDRMFrameDescriptor *)src->planes[0];
+
+    for (layer_no = 0; layer_no < avdesc->nb_layers; layer_no++) {
+        AVDRMLayerDescriptor layer = avdesc->layers[layer_no];
+
+        entry->drm_format = layer.format;
+        for (plane_no = 0; plane_no < layer.nb_planes; ++plane_no) {
+            AVDRMPlaneDescriptor plane = layer.planes[plane_no];
+            int object_index = plane.object_index;
+            AVDRMObjectDescriptor object = avdesc->objects[object_index];
+            uint64_t modifier = object.format_modifier;
+
+            zwp_linux_buffer_params_v1_add(params, object.fd, plane_no, plane.offset, plane.pitch, modifier >> 32, modifier & 0xffffffff);
+        }
+    }
+
+    return true;
+}
+
+static void resize(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wl;
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_osd_res osd;
+    const int width = wl->scaling * mp_rect_w(wl->geometry);
+    const int height = wl->scaling * mp_rect_h(wl->geometry);
+    
+    vo_wayland_set_opaque_region(wl, 0);
+    vo->dwidth = width;
+    vo->dheight = height;
+    vo_get_src_dst_rects(vo, &src, &dst, &osd);
+
+   if (wl->viewport)
+        wp_viewport_set_destination(wl->viewport, (dst.x0 << 1) + mp_rect_w(dst), (dst.y0 << 1) + mp_rect_h(dst));
+
+   if (wl->video_viewport)
+        wp_viewport_set_destination(wl->video_viewport, mp_rect_w(dst), mp_rect_h(dst));
+    wl_subsurface_set_position(wl->video_subsurface, dst.x0, dst.y0);
+    vo->want_redraw = true;
+}
+
+static void draw_frame(struct vo *vo, struct vo_frame *frame)
+{
+    struct priv *p = vo->priv;
+    struct vo_wayland_state *wl = vo->wl;
+    struct wlbuf_pool_entry *entry;
+    struct mp_image *current;
+    
+    if (!vo_wayland_check_visible(vo))
+        return;
+
+    current = mp_image_new_ref(frame->current);
```

Comment:
Doesn't this technically make a copy of every frame? I'm not sure how much it matters in practice but `vo_vaapi_wayland` doesn't do this. I vaguely recall there were issues though.

Reply:
Cool, makes sense.